File: kotlin-node/karakum/annotations/annotateHttpsServer.js
import ts from "typescript"

export default (node) => {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

    if (
        sourceFileName.endsWith("https.d.ts")
        && ts.isClassDeclaration(node)
        && node.name?.text === "Server"
    ) {
        return `@Suppress("MANY_CLASSES_IN_SUPERTYPE_LIST")`
    }

    if (
        sourceFileName.endsWith("https.d.ts")
        && ts.isExpressionWithTypeArguments(node)
        && ts.isPropertyAccessExpression(node.expression)
        && ts.isIdentifier(node.expression.expression)
        && node.expression.expression.text === "http"
        && ts.isIdentifier(node.expression.name)
        && node.expression.name.text === "Server"

        && node.parent
        && ts.isHeritageClause(node.parent)

        && node.parent.parent
        && ts.isInterfaceDeclaration(node.parent.parent)
        && node.parent.parent.name?.text === "Server"
    ) {
        return `@seskar.js.JsMixin`
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/wrappersgenerator/node/annotations/annotateHttpsServer.kt
package wrappersgenerator.node.annotations

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.AnnotationContext
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.*

fun annotateHttpsServer(node: Node, context: AnnotationContext) = nullable {
    val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName

    ensure(sourceFileName.endsWith("https.d.ts"))

    nullable {
        ensure(isClassDeclaration(node))
        ensure(node.name?.text == "Server")

        "@Suppress(\"MANY_CLASSES_IN_SUPERTYPE_LIST\")"
    } ?: nullable {
        ensure(isExpressionWithTypeArguments(node))

        val expression = node.expression
        ensure(isPropertyAccessExpression(expression))

        val expressionExpression = expression.expression
        ensure(isIdentifier(expressionExpression))
        ensure(expressionExpression.text == "http")

        val expressionName = expression.name
        ensure(isIdentifier(expressionName))
        ensure(expressionName.text == "Server")

        val heritageClause = ensureNotNull(node.getParentOrNull())
        ensure(isHeritageClause(heritageClause))

        val interfaceNode = ensureNotNull(heritageClause.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(interfaceNode.name.text == "Server")

        "@seskar.js.JsMixin"
    }
}
